frame/contracts/src/lib.rs

```diff
@@ -913,9 +917,33 @@ pub mod pallet {
 		StorageValue<_, DeletionQueueManager<T>, ValueQuery>;
 }
 
+/// The type of origins supported by the contracts pallet.
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo, RuntimeDebugNoBound)]
+pub enum Origin<T: Config> {
+	Root,
+	Signed(T::AccountId),
+}
+
+impl<T: Config> Origin<T> {
+	/// Creates a new Signed Caller from an AccountId.
+	pub fn from_account_id(account_id: T::AccountId) -> Self {
+		Origin::Signed(account_id)
+	}
+}
+
+impl<T: Config> Origin<T> {
+	/// Returns the account id of the caller if it has one.
+	/// It errors otherwise.
+	pub fn account_id(&self) -> Result<&T::AccountId, DispatchError> {
+		match self {
+			Origin::Signed(id) => Ok(id),
+			Origin::Root => Err(DispatchError::RootNotAllowed),
+		}
+	}
+}
+
 /// Context of a contract invocation.
 struct CommonInput<'a, T: Config> {
-	origin: T::AccountId,
```

Comment:
Let's maybe keep it the same for call and instantiate? Let it be `Origin`and then handle the Root case exlicitly in the instantiate case. I see this is not currently possible as `Origin::Root` only comes from the `call` dispatchable. Still, I think this would make code more clear.

Reply:
While I agree that it looks more consistent to have the same origin type on both `CallInput` and `InstantiateInput` (in which case we should actually move it to `CommonInput`), I would actually find it more confusing to have a struct that accepts a type that we should always check for a situation that it isn't even possible.
Other than adding the overhead of having to check for the Root case, it wouldn't be immediate obvious that the origin of an instantiate is always a plain account.

Reply:
> a type that we should always check for a situation that it isn't even possible.


What if we'll have to change the rules of which dispatchables which origin is allowed to call? Or if we'd later decide to add support for other origins?

Wouldn't putting these checks into `Invokable` trait implementations make our life easier in the future?

Reply:
Not all dispatchables that need to `ensure_signed` call `Invokable::run_guarded()`, `remove_code` and `upload_code` don't. While I like your suggestion I am not convinced of applying that generalization to only some dispatchables, then it is just a "partial generalization" (if that is even something).

Reply:
As discussed in DM, the ensure origin logic has been moved inside run guarded for those dispatchables that have an Invokable